scripts/check-type-dependencies.js

```diff
@@ -0,0 +1,194 @@
+        if (error.name === 'WrongDepError') {
+          console.error(
+            `  Move from ${chalk.red(error.from)} to ${chalk.green(
+              error.to,
+            )}: ${chalk.cyan(error.dep)}`,
+          );
+        } else if (error.name === 'MissingDepError') {
+          console.error(
+            `  Missing a type dependency: ${chalk.cyan(error.dep)}`,
+          );
+        } else {
+          console.error(`  Unknown error, ${chalk.red(error)}`);
+        }
+      }
+    }
+  }
+
+  if (hadErrors) {
+    console.error();
+    console.error(
+      chalk.red('At least one package had incorrect type dependencies'),
+    );
+
+    process.exit(2);
+  }
+}
+
+function shouldCheckTypes(pkg) {
+  return !pkg.private && pkg.get('types');
+}
+
+/**
+ * Scan index.d.ts for imports and return errors for any dependency that's
+ * missing or incorrect in package.json
+ */
+function checkTypes(pkg) {
+  const typeDecl = fs.readFileSync(
+    resolvePath(pkg.location, 'dist/index.d.ts'),
+    'utf8',
+  );
+  const deps = (typeDecl.match(/from '.*'/g) || [])
+    .map(match => match.replace(/from '(.*)'/, '$1'))
+    .filter(n => !n.startsWith('.'));
+
+  const errors = [];
+  const typeDeps = [];
+  for (const dep of deps) {
+    try {
+      const typeDep = findTypesPackage(dep, pkg);
+      if (typeDep) {
+        typeDeps.push(typeDep);
+      }
+    } catch (error) {
+      errors.push(error);
+    }
+  }
+
+  errors.push(...findTypeDepErrors(typeDeps, pkg));
+
+  return { errors };
+}
+
+/**
+ * Find the package used for types. This assumes that types are working is a package
+ * can be resolved, it doesn't do any checking of presence of types inside the dep.
+ */
+function findTypesPackage(dep, pkg) {
+  try {
+    require.resolve(`@types/${dep}/package.json`, { paths: [pkg.location] });
+    return `@types/${dep}`;
+  } catch {
+    try {
+      require.resolve(dep, { paths: [pkg.location] });
+      return undefined;
+    } catch {
+      try {
+        // Some type-only modules don't have a working main field, so try resolving package.json too
+        require.resolve(`${dep}/package.json`, { paths: [pkg.location] });
+        return undefined;
+      } catch {
+        try {
+          // Finally check if it's just a .d.ts file
+          require.resolve(`${dep}.d.ts`, { paths: [pkg.location] });
+          return undefined;
+        } catch {
+          throw mkErr('MissingDepError', `No types for ${dep}`, { dep });
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Figures out what type dependencies are missing, or should be moved between dep types
+ */
+function findTypeDepErrors(typeDeps, pkg) {
+  const devDeps = mkTypeDepSet(pkg.get('devDependencies'));
+  const deps = mkTypeDepSet(pkg.get('dependencies'));
+
+  const errors = [];
+  for (const typeDep of typeDeps) {
+    if (!deps.has(typeDep)) {
+      if (devDeps.has(typeDep)) {
+        errors.push(
+          mkErr('WrongDepError', `Should be dep ${typeDep}`, {
```

Comment:
could afford to write out as "Should be placed in dependencies instead of devDependencies"